drafts/str4d-overwinter-activation/draft1.rst

```diff
@@ -0,0 +1,229 @@
+
+
+Specification
+=============
+
+The following constants are defined for every network upgrade:
+
+BRANCH_ID
+  A globally-unique non-zero 32-bit identifier.
+
+  Implementations MAY use a value of zero in branch ID fields to indicate the absence of any upgrade (i.e.
+  that the Sprout consensus rules apply).
+
+ACTIVATION_HEIGHT
+  The non-zero block height at which the network upgrade rules will come into effect, and be enforced as part
+  of the blockchain consensus.
+
+  For removal of ambiguity, the block at height ``ACTIVATION_HEIGHT - 1`` is subject to the pre-upgrade
+  consensus rules, and would be the last common block in the event of a persistent pre-upgrade branch.
+
+  It MUST be greater than the value of ``DEPRECATION_HEIGHT`` in the last software version that will not
+  contain support for the network upgrade. It SHOULD be chosen to be reached approximately three months after
+  the first software version containing support for the network upgrade is released, for the following reason:
+
+  - As of the time of writing (the 1.0.15 release), the release cycle is six weeks long, and nodes undergo
+    auto-senescence 16 weeks after release. Thus, if version ``X`` contains support for a network upgrade,
+    version ``X-1`` will be deprecated 10 weeks after the release of version ``X``, which is about 2.3 months.
+    A three-month window provides ample time for users to upgrade their nodes after auto-senescence, and
+    re-integrate into the network prior to activation of the network upgrade.
+
+The relationship between ``BRANCH_ID`` and ``ACTIVATION_HEIGHT`` is many-to-one: it is possible for many
+distinct branches to descend from the same parent block (and thus have the same ``ACTIVATION_HEIGHT``), but a
+specific branch can only have one parent block. Concretely, this means that if the ``ACTIVATION_HEIGHT`` of a
+network upgrade is changed for any reason (e.g. security vulnerabilities or consensus bugs are discovered),
+the ``BRANCH_ID`` MUST also be changed.
+
+Activation mechanism
+--------------------
+
+A blockchain is defined as invalid if, within the set of all network upgrades that have activated in the past
+(or will activate in future) on that chain, an ``ACTIVATION_HEIGHT`` is repeated. Note that this does not
+require ``ACTIVATION_HEIGHT`` to be globally unique, or even locally unique; multiple network upgrades can
+occur in parallel, as long as they are non-overlapping (only one will activate on any given chain).
+
+Concretely, this means that the Zcash blockchain is broken into "epochs" of block height intervals
+``[ACTIVATION_HEIGHT_N, ACTIVATION_HEIGHT_{N+1})`` (i.e. including ``ACTIVATION_HEIGHT_N`` and excluding
+``ACTIVATION_HEIGHT_{N+1}``), on which consensus rule sets are defined.
+
+Consensus rules themselves (and any network behavior or surrounding code that depends on them) MUST be gated
+by block height checks. For example:
```

Comment:
Wait, some of them should and some of them shouldn't (e.g. if a rule is associated with a transaction version enabled by an upgrade, it doesn't need to be &ndash;and shouldn't be&ndash; gated by a height check if acceptance of that tx version is already gated by a height check elsewhere). I suggest changing to:

> When a consensus rule depends on activation of a particular upgrade, its implementation (and that of any network behaviour or surrounding code that depends on it) MUST be gated by a block height check.